src/custom/prelude/array.rs

```diff
@@ -0,0 +1,638 @@
+        .into_pointer_value();
+    builder.build_store(array_ptr, array)?;
+    go(ptr)
+}
+
+pub fn emit_array_op<'c, H: HugrView>(
+    pcg: &impl PreludeCodegen,
+    ctx: &mut EmitFuncContext<'c, H>,
+    op: ArrayOp,
+    inputs: Vec<BasicValueEnum<'c>>,
+    outputs: RowPromise<'c>,
+) -> Result<()> {
+    let builder = ctx.builder();
+    let ts = ctx.typing_session();
+    let sig = op.clone().to_extension_op().unwrap().signature();
+    let ArrayOp {
+        def,
+        ref elem_ty,
+        size,
+    } = op;
+    let llvm_array_ty = pcg
+        .array_type(&ts, ts.llvm_type(elem_ty)?, size)
+        .as_basic_type_enum()
+        .into_array_type();
+    match def {
+        ArrayOpDef::new_array => {
+            let mut array_v = llvm_array_ty.get_undef();
+            for (i, v) in inputs.into_iter().enumerate() {
+                array_v = builder
+                    .build_insert_value(array_v, v, i as u32, "")?
+                    .into_array_value();
+            }
+            outputs.finish(builder, [array_v.as_basic_value_enum()])
+        }
+        ArrayOpDef::get => {
+            let [array_v, index_v] = inputs
+                .try_into()
+                .map_err(|_| anyhow!("ArrayOpDef::get expects two arguments"))?;
+            let array_v = array_v.into_array_value();
+            let index_v = index_v.into_int_value();
+            let elem = with_array_alloca(builder, array_v, |ptr| {
+                let elem_addr = unsafe { builder.build_gep(ptr, &[index_v], "")? };
+                builder.build_load(elem_addr, "")
+            })?;
+
+            let res_sum_ty = {
+                let TypeEnum::Sum(st) = sig
+                    .output()
+                    .get(0)
+                    .ok_or(anyhow!("ArrayOp::get has no outputs"))?
+                    .as_type_enum()
+                else {
+                    Err(anyhow!("ArrayOp::get output is not a sum type"))?
+                };
+                LLVMSumType::try_new(&ts, st.clone())?
+            };
+            let success_v = res_sum_ty.build_tag(builder, 1, vec![elem])?;
+            let failure_v = res_sum_ty.build_tag(builder, 0, vec![])?;
+            let is_success = builder.build_int_compare(
+                IntPredicate::ULT,
+                index_v,
+                index_v.get_type().const_int(size, false),
+                "",
+            )?;
+            let res_v = builder.build_select(is_success, success_v, failure_v, "")?;
+            outputs.finish(ctx.builder(), [res_v])
+        }
+        ArrayOpDef::set => {
+            let [array_v0, index_v, value_v] = inputs
+                .try_into()
+                .map_err(|_| anyhow!("ArrayOpDef::set expects three arguments"))?;
+            let array_v = array_v0.into_array_value();
+            let index_v = index_v.into_int_value();
+            let (elem_v, array_v) = with_array_alloca(builder, array_v, |ptr| {
```

Comment:
Performing the insertion, *then* checking whether the index is in bounds seems very dangerous. Am I missing something here?

Reply:
I think the same applies for swap